docs/guides/testing-keycloak.rst

```diff
@@ -0,0 +1,111 @@
+
+.. code-block:: yaml
+
+    domains:
+    - id: sssd
+      hosts:
+      - hostname: idm.ipa.test
+        role: ipa
+        ssh:
+          host: HOSTNAME_IPA
+          username: root
+          password: Secret123
+        config:
+          client:
+          ipa_domain: ipa.test
+
+      - hostname: bridge.ipa.test
+        role: client
+        ssh:
+          host: HOSTNAME_BRIDGE
+          username: root
+          password: Secret123
+
+      - hostname: keycloak.ipa.test
+        role: keycloak
+        ssh:
+          host: HOSTNAME_KEYCLOAK
+          username: root
+          password: Secret123
+        config:
+          keycloak:
+            keycloak_server: keycloak.ipa.test
+            keycloak_admin_user: admin
+            keycloak_admin_password: Secret123
+
+It should be noted that at this time, the ipa-tuura bridge service does not
+yet have its own role.  Instead it is defined as a client for remote access if
+necessary and is managed via the Keycloak SCIM storage plugin.
+
+Adding a Keycloak Integration Domain
+------------------------------------
+
+Keycloak integration requires the Keycloak server to add an Integration Domain
+that will point to the identity provider being tested (IPA, AD, LDAP, etc).
+This is accomplished using the method
+:meth:`sssd_test_framework.roles.keycloak.KeycloakDomain.add`.   After Adding
+the Integration Domain, you can add users to keycloak and they should be synced
+to the identity provider.  At that point, you should be able to use them like
+any user created for that provider.
+
+Test example
+============
+
+.. code-block:: python
+
+    from __future__ import annotations
+
+    import pytest
+
+    from pytest_mh import Topology, TopologyDomain
+
+    from sssd_test_framework.roles.client import Client
+    from sssd_test_framework.roles.keycloak import Keycloak
+    from sssd_test_framework.roles.ipa import IPA
+    from sssd_test_framework.roles.generic import GenericProvider
+
+
+    @pytest.mark.topology(
+        "ipa",
+        Topology(TopologyDomain("sssd", client=1, keycloak=1, ipa=1)),
+        keycloak="sssd.keycloak[0]", client="sssd.client[0]", ipa="sssd.ipa[0]", provider="sssd.ipa[0]"
+    )
+    def test_keycloak__lookup_keycloak_user_in_domain(client: Client, keycloak: Keycloak, provider: GenericProvider):
+        """
+        :title: Lookup Keycloak user in domain
+        :steps:
+            1. add user to keycloak
+            2. lookup user in domain
+        :expectedresults:
+            1. null
+            2. user found in domain
+        """
+        keycloak.domain("scimipa").add(
+            scim_host="bridge.ipa.test",
+            scim_admin="scim",
+            scim_password="Secret123",
+            dom_name="ipa.test",
+            dom_url="http://ipa.test",
+            client_id="admin",
+            client_secret="Secret123",
+            id_provider="ipa",
+            ldap_tls_cacert="/etc/ipaca.crt",
+        )
```

Comment:
This would be good to add to the setup of the keycloak role.

Or maybe even better, it should be made a prerequisite for the host. Similar what we have on client - client must be enrolled into domains, so we don't need to reenroll it every time.

What do you think?

Reply:
Another thing I was going to ask about.

Keycloak.domain.add() should (almost) always connect with the info from provider (IPA|AD|LDAP).    How could I accomplish this?  In the definition of the KnownTopology somehow?   

Or is it easier to just define that info as new config attributes in the mhc.yaml?   That might also eliminate the need for a bridge role in the mhc.yaml (until we actually need one (if ever)).

Reply:
I realized that I can't define the info needed in the config under bridge or keycloak because it has to come from the parametrized provider.   Is there an easier way to do that in the Keycloak.setup to run a self.domain.add() than to use something like:

```
for p in ['ipa', 'ad', 'ldap']:
    if self.host.domain.hosts_by_role(p):
        break
if not p:
    raiseSomeError

provider = host.domain.hosts_by_role(p)[0]

self.domain.add(
...
client_secret=provider.adminpw
...)
```

Is there a shorter/easier way to find the provider currently in use?

Reply:
It depends on what you want to. Is the bridge connected to all backends or always just to one? You can put it to mhc.yaml but you may need to have more configurations.

But if you look what we do with "client", we add sssd.conf options from the provider host's mhc config. This allows us to configure SSSD based on the currently selected topology.

We extend the pytest-mh topology with custom stuff that allows us to do that, see:
* https://github.com/SSSD/sssd-test-framework/blob/master/sssd_test_framework/topology.py#L17
* https://github.com/SSSD/sssd-test-framework/blob/master/sssd_test_framework/config.py#L17

I would say don't try to montage it all to SSSD project requirements. ipa-tuura requires slightly different things, so maybe you need to make your extensions to allow to test it neatly.

It may be good to have a call and see what needs to be done.